File: Linked_List/Java/Linked List in Binary Tree.java
// Problem link: https://leetcode.com/problems/linked-list-in-binary-tree/description
// Time: O(n * m)
// Space: O(n + m)
class Solution {
    public boolean isSubPath(ListNode head, TreeNode root) {
        if(root == null) {
            return false;
        }
        return this.check(head, root) || this.isSubPath(head, root.left) || this.isSubPath(head, root.right);
    }

    private boolean check(ListNode head, TreeNode root) {
        if(head == null) {
            return true;
        }
        else if(root == null || head.val != root.val) {
            return false;
        }
        return this.check(head.next, root.left) || this.check(head.next, root.right);
    }
}

File: Linked_List/Python/Linked List in Binary Tree.py
# Problem link: https://leetcode.com/problems/linked-list-in-binary-tree/description
# Time: O(n * m)
# Space: O(n + m)
class Solution:
    def isSubPath(self, head: Optional[ListNode], root: Optional[TreeNode]) -> bool:
        if not root:
            return False
        return self.check(head, root) or self.isSubPath(head, root.left) or self.isSubPath(head, root.right)
        
    def check(self, listNode, treeNode):
        if not listNode:
            return True
        elif not treeNode or listNode.val != treeNode.val:
            return False
        
        return self.check(listNode.next, treeNode.left) or self.check(listNode.next, treeNode.right)
